docs/design/resource-qos.md

```diff
@@ -0,0 +1,247 @@
+
+Pods can be of one of 3 different classes:
+
+- If `limits` and optionally `requests` (not equal to `0`) are set for all resources across all containers and they are *equal*, then the container is classified as **Guaranteed**.
+
+Examples:
+
+```yaml
+containers:
+	name: foo
+		resources:
+			limits:
+				cpu: 10m
+				memory: 1Gi
+	name: bar
+		resources:
+			limits:
+				cpu: 100m
+				memory: 100Mi
+```
+
+```yaml
+containers:
+	name: foo
+		resources:
+			limits:
+				cpu: 10m
+				memory: 1Gi
+			requests:
+				cpu: 10m
+				memory: 1Gi
+
+	name: bar
+		resources:
+			limits:
+				cpu: 100m
+				memory: 100Mi
+			requests:
+				cpu: 10m
+				memory: 1Gi
+```
+
+- If `requests` and optionally `limits` are set (not equal to `0`) for one or more resources across one or more containers, and they are *not equal*, then the pod is classified as **Burstable**.
+When `limits` are not specified, they default to the node capacity.
+
+Examples:
+
+Container `bar` has not resources specified.
+
+```yaml
+containers:
+	name: foo
+		resources:
+			limits:
+				cpu: 10m
+				memory: 1Gi
+			requests:
+				cpu: 10m
+				memory: 1Gi
+
+	name: bar
+```
+
+Container `foo` and `bar` have limits set for different resources.
+
+```yaml
+containers:
+	name: foo
+		resources:
+			limits:
+				memory: 1Gi
+
+	name: bar
+		resources:
+			limits:
+				cpu: 100m
+```
+
+Container `foo` has no limits set, and `bar` has neither requests nor limits specified.
+
+```yaml
+containers:
+	name: foo
+		resources:
+			requests:
+				cpu: 10m
+				memory: 1Gi
+
+	name: bar
+```
+
+- If `requests` and `limits` are not set for all of the resources, across all containers, then the pod is classified as **Best-Effort**.
+
+Examples:
+
+```yaml
+containers:
+	name: foo
+		resources:
+	name: bar
+		resources:
+```
+
+Pods will not be killed if CPU guarantees cannot be met (for example if system tasks or daemons take up lots of CPU), they will be temporarily throttled.
+
+Memory is an incompressible resource and so let's discuss the semantics of memory management a bit.
+
+- *Best-Effort* pods will be treated as lowest priority. Processes in these pods are the first to get killed if the system runs out of memory.
+These containers can use any amount of free memory in the node though.
+
+- *Guaranteed* pods are considered top-priority and are guaranteed to not be killed until they exceed their limits, or if the system is under memory pressure and there are no lower priority containers that can be evicted.
+
+- *Burstable* pods have some form of minimal resource guarantee, but can use more resources when available.
+Under system memory pressure, these containers are more likely to be killed once they exceed their requests and no *Best-Effort* pods exist.
+
+### OOM Score configuration at the Nodes
+
+Pod OOM score configuration
```

Comment:
I am ok with this change today to applying one oom score for entire pod. 

But in the future, I am thinking distinguishing the containers within a pod with a different oom score? For example two containers in a Pod: A and B, A has both request and limit specified and they are equal, B doesn't have `requests` and `limits` being. With the new policy introduced in this pr, the pod is burstable, but the user might want to oom kill to kill B first before A. I know there are a lot of corner cases to resolve, let's iterate this later.